### DISCUSSION TITLE
Moving photos that are in between certain parameters automatically

### DISCUSSION THREAD

**Discussion by smpetit** [ImageMagick/ImageMagick#4007] [Help]
Hi there!  I apologize if I'm not posting in the right area (any pointers there would be helpful!), but I'm hoping to gain some insight on the proper methods I can use to organize some files automatically.

I have attached a photo for reference to what I am trying to describe here.

I take photographs at equestrian sports for competitors, and we can shoot over 150 different people a day riding their horses. At the end of a "Set" (a group of about 15 people), they switch Sets, and we start shooting new people. Essentially I would like to move/copy the files from an SD card of a Set, to the indicated folders "1, 2, 3, 4, etc." to then run some script I have made to then move the files from the first folder ("1"), to their designated rider folder (i.e, "Billy Bob").  

I'm not sure of the direction to take here, but my thought would be that ImageMagick could detect a "Black" photo (What we call a Dirt Shot), then move the files in between the Dirt Shots to the corresponding folder.  Referencing the photo - so the first shots in between the two dirt shots would go to Folder 1, the next to Folder 2, etc.

The image sizes for the Dirt Shots are roughly 9MB to 9.6MB.  The regular shots are anywhere between 17MB and 21MB.

I am using Mac OS Catalina and the latest version of ImageMagick to organize my photos currently, but will eventually be using a Mac Mini with Big Sur - if that makes any difference.

Does anyone have any ideas on how to script this??  Or is it even doable?  

I appreciate anyone and everyone's time and thoughts on this project!

-Seth

<img width="428" alt="Screen Shot 2021-07-31 at 10 21 08 AM" src="https://user-images.githubusercontent.com/88241514/127744554-1f0b36f6-6888-489e-bf4b-5bd35b2de205.png">

**Comment by fmw42** [comment]
You can use a bash shell script to loop over each image. ImageMagick can be scripted in the loop to measure the average brightness (mean). If the mean is above some threshold move to folder 1. If the mean is below some threshold, skip the black image and increment the folder to 2. Then go to the next image and continue the process. This is mostly bash shell scripting. The only part that is imagemagick is getting the average brightness, which would be in a binary test (0 or 1)

test=`convert image.jpg -format "%[fx:mean<$threshold?1:0]" info:`

if test=1 it was below threshold, so increment folder and go to next image
if test=0 it was equal or above threshold, so move image to current folder

**Reply by smpetit** [reply]
I will dig some more and see if I can get a working model using your idea!  Thank you for the input!  I'm still extremely "green" when it comes to command-line scripting.

So essentially, after the test comes back, would I then put the rest of the script to designate to move to Folder 1, etc.?

**Reply by fmw42** [reply]
The move (if you really want to do that rather than a copy) would be in the conditional testing for a 0 or 1 from test

```
test=$(convert image.jpg -format "%[fx:mean<$threshold?1:0]" info:)
if [ $test -eq 0 ]; then
   # not black
   mv current_directory/file new_directory/file
elif [ $test -eq 1]; then
   # black
   new_directory_id = current_directory_id + 1
fi

```

**Reply by smpetit** [reply]
Hi again!  
Tried just a quick test, but I'm probably overlooking something to make it look at every file in the "fake" SD card for the test.

I tried changing "image.jpg" to " *.jpg " and it didn't seem to like that very much.

The Mac Terminal said this:
                                            zsh: command not found: image.JPG
                                            [: unknown condition: -eq
                                            [: '] ' expected

//
test=convert image.jpg -format "%[fx:mean<$threshold?1:0]" info:
if [ $test -eq 0 ]; then
   # not black
   mv /Users/s______/Desktop/SD_CARD /Users/s______/Desktop/MASTER/1
elif [ $test -eq 1]; then
   # black
   /Users/s______/Desktop/MASTER/1 = /Users/s_____/Desktop/MASTER/1 + 1
fi

//
Sorry.. Not sure how to paste code nice and neat like you did!

**Reply by fmw42** [reply]
image.jpg is whatever your files are name. 

You have to loop over each file in the folder. This is bash not zsh. So you may have to change shells. It is also untested, so be sure you have backups or change mv to cp to copy.  

Please read about bash scripting. This is an ImageMagick forum and not shell scripting forum. We are happy to help. But this is mostly shell scripting and we can only help so much on that.

```
i=1
threshold=enter some number between 0 and 1. #Threshold=0 would be pure black and 1 would be pure white.
cd your_image_folder
for img in *.jpg; do
echo "$img"
test=$(convert "$img" -format "%[fx:mean<$threshold?1:0]" info:)
if [ $test -eq 0 ]; then
echo "black"
i=$((i+1))
elif [ $test -eq 1 ]; then
echo "not black"
mv $img path_to_folder_$i/$img
fi

```
I suggest you separate a small list of files to a new directory and work on that to be sure the script is working.

**Reply by fmw42** [reply]
See if then conditionals at https://ryanstutorials.net/bash-scripting-tutorial/bash-if-statements.php Also see for loops at https://www.cyberciti.biz/faq/bash-for-loop/

**Reply by snibgo** [reply]
fmw42 wrote:
```
test=convert image.jpg -format "%[fx:mean<$threshold?1:0]" info:
```
I am no expert, but shouldn't there be backticks or something?

**Reply by fmw42** [reply]
Yes, thanks. I am not sure what happened. Thought I had put them, but must have overlooked it.

**Reply by smpetit** [reply]
I really appreciate the insight!  Thank you for the links to the information!

**Reply by fmw42** [reply]
Let us know if it works or if there is an error and we can try to help to fix it. Please test first on a small set.

**Reply by smpetit** [reply]
Hi there!  I got it to finally not have any errors, but it labeled all the files as black and just kept skipping them.  haha  (And it seemed to take a while to run through them)

I found an old thread that I believe you responded to around 10 years ago, where someone used a command to find the file size and organize their photos that way.  I did some testing, and I also couldn't quite get it to work.. but it at least moved the files over -  it just never detected the "black" files to skip to the new folder.  Effectively, all of the "black" shots are well under 15MB, while the actual photos are around the 17-21MB range. (I was just using 15MB as a baseline since I knew the test files met that criteria)

Here is the script I used for that:
#!/bin/bash
cd /Users/s______/Desktop/SD_CARD
i=1
for img in *.JPG;
do
echo "$img"
test=$(identify -format "%b" "$img")
if [ "15MB" < $test ]; then
echo "black"
i=$((i+1))
elif [ "15MB" > $test ]; then
echo "not black"
mv $img /Users/s_____Desktop/MASTER/$i/$img
fi
done

I'll keep plugging along and trying to find out where it's not going quite right.  
-Thank you again though!  You have pointed me in the right direction.

**Reply by snibgo** [reply]
>  ... but it labeled all the files as black and just kept skipping them. haha

What value did you set for $threshold ?

**Reply by smpetit** [reply]
I tried 0 for black.. then tried 1 just to see if there was any change, and that definitely didn’t work at all.

**Reply by fmw42** [reply]
You need to set a larger threshold for my script. Check the value for the mean for you black images.

`convert black_image -format "%[fx:mean]" info:
`

what does it return when you test on a number of your black images?  Then make it a bit larger.

If your black image is JPG, then it may not be completely back due to JPG compression. Also if you just scanned a black image, it won't be completely black.

Your test

`if [ "15MB" < $test ]; then
`

will not work, because bash does not test for characters. You need to strip the MB etc from the value and test without it use [15 -lt $test]. You will be better to use 

%B rather than %b

%B returns the value in bytes without the B. 

See https://imagemagick.org/script/escape.php

**Reply by fmw42** [reply]
Please show your exact code that you used with my script that you say does not move any files? Did you get any error messages?

**Comment by snibgo** [comment]
As fmw42 says.

> Essentially I would like to move/copy the files from an SD card of a Set, to the indicated folders "1, 2, 3, 4, etc." ...

Personally, I think doing any processing on files on a SD card, other than a simple and direct copying of files, is unwise. When something goes wrong, you lose the only copy of the image. For an example, see the recent thread https://github.com/ImageMagick/ImageMagick/discussions/3998.

**Reply by smpetit** [reply]
Hi there!  I'm sorry I didn't put the entire process in my post!  
I make a backup of the card before I transfer anything to anywhere - that way in case one of my staff members messes up a folder, we can go back to the original card.   Thank you for the note though!

**Comment by fmw42** [comment]
I am sorry, I said move. But I really meant copy. Also use the file system to do the copy. But if you have a backup, then you could use the file system to move the file. Do not use ImageMagick to copy the file as it will read, decompress, recompress and write the new file. So you could have compression loss if you are using a lossy compressed format.

**Reply by fmw42** [reply]
Apparently backticks do not show in this forum (in one line code blocks) as they conflict with the backticks used for marking code the one line code blocks. So I have changed them to $(....).

**Comment by fmw42** [comment]
My code had some issues. I forgot the ending "done" to end the for loop. I had the test reversed. So the corrected script would be (assuming a threshold of 0.1):

```
cd your_image_folder
i=1
threshold=0.1
for img in *.jpg; do
echo "$img"
test=$(convert "$img" -format "%[fx:mean<$threshold?1:0]" info:)
if [ $test -eq 1 ]; then
echo "black"
i=$((i+1))
elif [ $test -eq 0 ]; then
echo "not black"
mv $img path_to_folder_$i/$img
fi
done

```
